Added ExperienceHandler and async network update algorithms

aboulang2002 opened the pull request:
Signed-off-by: Alexandre Boulanger <aboulang2002@yahoo.com>

## What changes were proposed in this pull request?

The goal of this PR is to externalize experience handling and network update from the learning classes. A few experience handling classes has been added in the package `org.deeplearning4j.rl4j.experience` and the network update logic has been extracted from `A3CThreadDiscrete` and `AsyncNStepQLearningDiscrete` to `A3CUpdateAlgorithm` and `QLearningUpdateAlgorithm` respectively.

I'd like to merge the sync and async update algorithms into a common interface (sync outputs a dataset with feature-labels and async outputs Gradient[]). To be done later.
Also note that the A3C update algorithm is not fully unit tested; I don't think its implementation is correct.

## How was this patch tested?

unit tests and manual tests

## Quick checklist

The following checklist helps ensure your PR is complete:

- [X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [X] Created tests for any significant new code additions.
- [X] Relevant tests for your changes are passing.

Bam4d commented on rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/experience/StateActionExperienceHandler.java:
```diff
@@ -0,0 +1,62 @@
+     * @return The list of experience elements
+     */
+    @Override
+    public List<StateActionPair<A>> getExperience() {
```
This API is pretty confusing as it looks like it should just get a list of the experience, but it also deletes all the current experience in the handler. 
I spent ages trying to debug something and found out this was the cause.

aboulang2002 replied:
Sorry for wasting your time. The worst is that I started changing this design and reverted it to the current one saying to myself "how experience is handled internally shouldn't be exposed".

I will change it.

Bam4d replied:
Haha no need to apologise.

Bam4d commented on rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/experience/ReplayMemoryExperienceHandler.java:
```diff
@@ -0,0 +1,75 @@
+     * @return A batch of experience selected from the replay memory. The replay memory is unchanged after the call.
+     */
+    @Override
+    public List<Transition<A>> getExperience() {
```
Again this is a bit confusing as it looks like a getter.

perhaps there should be "sampleExperience" function

aboulang2002 replied:
The thing is "sampleExperience" is correct only in the context of replay memory. But I understand what you mean. I'll try to find something better.

Bam4d replied:
There are different types of replay memory (e.g hindsight, prioritised etc) so it might be good to have an interface for it that has "sampleExperience".

If you are using an experience replay algorithm a common e.g "sampleExperience(int batchSize)" function would be mean you can plug and play different impls.

aboulang2002 replied:
I disagree. A sample is only used in memory replay types of algorithms, it makes no sense for example with A3C where a slice of the trajectory is used. I want any class implementing `ExperienceHandler` to be able to be used by any actor-learner, having a "sampleExperience(int batchSize)" would prevent that.

aboulang2002 replied:
I changed the name of the method.  With `generateTrainingBatch()`, users shouldn't have an expectation of idempotency, contrary to `getExperience()`.

aboulang2002 commented:
Summary of changes:
 - fixed a problem where a local var needed to be marked as final.
 - Added a builder pattern to `ReplayMemoryExperienceHandler`
 - Changed the method name from `getExperience()` to `generateTrainingBatch()`

aboulang2002 commented:
Removed this PR because of unfixable bad merge
